File: apps/android/app/src/main/java/dev/flextranslate/foundation/CloudCallGate.kt
package dev.flextranslate.foundation

/**
 * Жёсткий гейт, который каждый облачный MT-вызов обязан пройти ДО любого сетевого трафика
 * (WS5 §2.4 / §3).
 *
 * Оборачивает существующие предусловия [CloudOptInState.canStart] — явное согласие пользователя,
 * принятое раскрытие данных и наличие сети — и превращает отказ в причину [Decision.Blocked] на
 * продуктовом языке. Причины повторяют список «чего не хватает» с `CloudScreen`, чтобы UI и
 * провайдер говорили ОДНУ И ТУ ЖЕ правду.
 *
 * Поддерживаются два режима учётных данных (см. [GeminiCredentialMode]):
 *
 * **BACKEND_MEDIATION** (исходный путь): дополнительно нужны настроенный backend-endpoint и живой
 * эфемерный токен, выданный бэкендом. Ключ Gemini не покидает сервер.
 *
 * **OWN_KEY** (BYOK): пропускает проверки бэкенда/токена — свой зашифрованный ключ пользователя
 * передаёт вызывающий код ([GeminiKeyStore]). Согласие, раскрытие и сеть всё равно обязательны.
 * Ключ здесь НИКОГДА не читаем, не храним и не логируем.
 */
class CloudCallGate(
    private val stateProvider: (providerId: String) -> CloudOptInState?,
    private val config: GeminiFlashConfig,
    /** Передаётся опционально для режима OWN_KEY, чтобы гейт мог проверить наличие ключа. */
    private val keyStore: GeminiKeyStore? = null,
) {
    fun evaluate(providerId: String, nowEpochMs: Long): Decision {
        val state = stateProvider(providerId)
            ?: return Decision.Blocked(REASON_DISABLED)

        if (!state.userConsented) return Decision.Blocked(REASON_NO_CONSENT)
        if (!state.disclosureAccepted) return Decision.Blocked(REASON_NO_DISCLOSURE)
        if (state.networkState != "online") return Decision.Blocked(REASON_OFFLINE)

        return when (config.credentialMode) {
            GeminiCredentialMode.OWN_KEY -> {
                // Бэкенд и эфемерный токен не проверяем. Но ключ должен быть.
                if (keyStore?.hasKey() != true) return Decision.Blocked(REASON_NO_OWN_KEY)
                // Для OWN_KEY подсовываем заглушку-credential, чтобы sealed-тип остался единым.
                Decision.Allowed(CloudCredential(source = "own_key", expiresAtEpochMs = Long.MAX_VALUE))
            }
            GeminiCredentialMode.BACKEND_MEDIATION -> {
                // Настроенный backend-endpoint — обязательное условие для медиированного пути.
                if (!config.hasBackend) return Decision.Blocked(REASON_NO_BACKEND)
                val credential = state.credential
                if (credential?.isEphemeral(nowEpochMs) != true) {
                    return Decision.Blocked(REASON_NO_TOKEN)
                }
                Decision.Allowed(credential)
            }
        }
    }

    sealed interface Decision {
        data class Allowed(val credential: CloudCredential) : Decision
        data class Blocked(val userReason: String) : Decision
    }

    companion object {
        const val REASON_NO_BACKEND = "Не указан backend-endpoint (Облако)"
        const val REASON_DISABLED = "Облако выключено"
        const val REASON_NO_CONSENT = "Облако выключено — нет согласия"
        const val REASON_NO_DISCLOSURE = "Нужно принять раскрытие данных"
        const val REASON_OFFLINE = "Нет сети — облачный перевод недоступен"
        const val REASON_NO_TOKEN = "Требуется backend-токен"
        const val REASON_NO_OWN_KEY = "Gemini API-ключ не задан"
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/CloudCallGate.swift
import Foundation

/// The hard gate every cloud MT call must pass BEFORE any network traffic.
///
/// Wraps the CloudOptInState preconditions — explicit user consent, accepted disclosure,
/// and online network — and turns a failure into a product-language Decision.blocked reason.
///
/// Two credential modes are supported (see GeminiCredentialMode):
///
/// backendMediation: also requires a configured backend endpoint and a live
/// backend-issued ephemeral token. The Gemini key never leaves the server.
///
/// ownKey (BYOK): skips backend/token checks — the user's own encrypted Keychain key is
/// supplied by the caller. Still requires consent, disclosure, and an online network.
/// The key is NEVER inspected, stored, or logged here.
///
/// Mirrors Android CloudCallGate.
final class CloudCallGate: @unchecked Sendable {

    private let stateProvider: @Sendable (String) -> CloudOptInState?
    private let config: GeminiFlashConfig
    private let keyStore: (any GeminiKeyStore)?

    init(
        stateProvider: @escaping @Sendable (String) -> CloudOptInState?,
        config: GeminiFlashConfig,
        keyStore: (any GeminiKeyStore)? = nil
    ) {
        self.stateProvider = stateProvider
        self.config = config
        self.keyStore = keyStore
    }

    enum Decision: Sendable {
        case allowed(CloudCredential)
        case blocked(String)
    }

    func evaluate(providerId: String, nowEpochMs: Int64) -> Decision {
        guard let state = stateProvider(providerId) else {
            return .blocked(Self.reasonDisabled)
        }
        guard state.userConsented else { return .blocked(Self.reasonNoConsent) }
        guard state.disclosureAccepted else { return .blocked(Self.reasonNoDisclosure) }
        guard state.networkState == "online" else { return .blocked(Self.reasonOffline) }

        switch config.credentialMode {
        case .ownKey:
            guard keyStore?.hasKey() == true else { return .blocked(Self.reasonNoOwnKey) }
            // Synthesise a dummy credential so the sealed type stays unified.
            return .allowed(CloudCredential(source: "own_key", expiresAtEpochMs: Int64.max))

        case .backendMediation:
            guard config.hasBackend else { return .blocked(Self.reasonNoBackend) }
            guard let credential = state.credential,
                  credential.isEphemeral(nowEpochMs: nowEpochMs) else {
                return .blocked(Self.reasonNoToken)
            }
            return .allowed(credential)
        }
    }

    // MARK: - Honest gate reasons (mirror Android constants)

    static let reasonNoBackend = "No backend endpoint configured (Cloud)"
    static let reasonDisabled = "Cloud is disabled"
    static let reasonNoConsent = "Cloud is disabled — no consent"
    static let reasonNoDisclosure = "Disclosure acceptance required"
    static let reasonOffline = "No network — cloud translation unavailable"
    static let reasonNoToken = "Backend token required"
    static let reasonNoOwnKey = "Gemini API key not set"
}
